### PR TITLE
feat(coderd): add template app usage to insights

### LINKED ISSUE #8658: Track `coder_app` usage in insights API
We're working on a user insights API (#8514, #8519, #8656) which can tell which IDEs developers are using per-template. We should also show what the popular `coder_app`s are (e.g. code-server, RStudio, JupyterLab, Matlab)

### PR DESCRIPTION
**Pull request by mafredri**
This PR adds template app usage to insights. Once merged, the dashboard will automatically start showing these.

Fixes #8658

Notable changes:

- Simplified the `generate_series` a bit via better `sqlc` workaround
- Query indentation was fixed (there was some tab spacing)
- Changed usage granularity from 5 minutes to 1 minute, I feel 5 minutes may skew the numbers too much for short lived sessions

Notes:

- Unfortunately I couldn't combine `GetTemplateInsights` and `GetTemplateAppInsights` into a single query because of `sqlc` limitations, and I did not want to trade type-safety for performance

### CODE REVIEW

**Comment by mafredri** [comment]
Demo:
image

File: coderd/database/queries/insights.sql [hunk 1]
```diff
@@ -22,18 +22,18 @@ GROUP BY workspace_agent_stats.user_id, users.username, users.avatar_url
...
 -- name: GetTemplateInsights :one
--- GetTemplateInsights has a granularity of 5 minutes where if a session/app was
--- in use, we will add 5 minutes to the total usage for that session (per user).
-WITH d AS (
-	-- Subtract 1 second from end_time to avoid including the next interval in the results.
-	SELECT generate_series(@start_time::timestamptz, (@end_time::timestamptz) - '1 second'::interval, '5 minute'::interval) AS d
-), ts AS (
+-- GetTemplateInsights has a granularity of 1 minute where if a session/app was
+-- in use during a minute, we will add 1 minute to the total usage for that
+-- session/app (per user).
```
Comment (mtojek) [inline coderd/database/queries/insights.sql:27]: Can we extract this into a separate PR? If we introduce a bug, it will be easier to spot the issue (possibly related to changed interval than app insights).

File: coderd/database/queries/insights.sql [hunk 2]
```diff
@@ -22,18 +22,18 @@ GROUP BY workspace_agent_stats.user_id, users.username, users.avatar_url
 ORDER BY user_id ASC;
 
 -- name: GetTemplateInsights :one
--- GetTemplateInsights has a granularity of 5 minutes where if a session/app was
--- in use, we will add 5 minutes to the total usage for that session (per user).
-WITH d AS (
-	-- Subtract 1 second from end_time to avoid including the next interval in the results.
-	SELECT generate_series(@start_time::timestamptz, (@end_time::timestamptz) - '1 second'::interval, '5 minute'::interval) AS d
-), ts AS (
+-- GetTemplateInsights has a granularity of 1 minute where if a session/app was
```
Comment (mtojek) [inline coderd/database/queries/insights.sql:25]: Do we need to adjust site/UI according to this change?

File: coderd/database/queries/insights.sql [hunk 3]
```diff
@@ -22,18 +22,18 @@ GROUP BY workspace_agent_stats.user_id, users.username, users.avatar_url
...
 	SELECT
 		d::timestamptz AS from_,
-		(d + '5 minute'::interval)::timestamptz AS to_,
-		EXTRACT(epoch FROM '5 minute'::interval) AS seconds
-	FROM d
-), usage_by_user AS (
+		(d::timestamptz + '1 minute'::interval) AS to_,
+		EXTRACT(epoch FROM '1 minute'::interval) AS seconds
+	FROM
+		-- Subtract 1 second from end_time to avoid including the next interval in the results.
```
Comment (mtojek) [inline coderd/database/queries/insights.sql:34]: Correct me if I'm wrong, but for 7 days, it will give 7 * 24 * 60 records per every JSON response? If there is really low usage, there will be a lot of empty slots. If this is a valid concern, maybe we should minimize the response.

File: coderd/insights.go [hunk 4]
```diff
@@ -215,6 +216,15 @@ func (api *API) insightsTemplates(rw http.ResponseWriter, r *http.Request) {
 			return xerrors.Errorf("get template insights: %w", err)
 		}
 
+		appUsage, err = tx.GetTemplateAppInsights(ctx, database.GetTemplateAppInsightsParams{
+			StartTime:   startTime,
+			EndTime:     endTime,
+			TemplateIDs: templateIDs,
```
Comment (mtojek) [inline coderd/insights.go:222]: Hint: while you are writing unit tests, please make sure that cover paths with `TemplateIDs` and without them (ALL).

File: coderd/database/queries/insights.sql (hunk 2 above)
Comment (mafredri) [reply coderd/database/queries/insights.sql:25]: Do you mean the change from 5m -> 1m? Nope, this is just a change in how much "time spent in X" is calculated. This may reduce time spent slightly due to less over-attribution.

File: coderd/database/queries/insights.sql (hunk 3 above)
Comment (mafredri) [reply coderd/database/queries/insights.sql:34]: It's a fair point, it will slow the query down a bit. I'd need to test against a real-ish dataset to figure out by how much. For very little data the overhead of 1m vs 5m is not too bad at ~65ms vs ~35ms, but that's just for one week.

File: coderd/database/queries/insights.sql (hunk 1 above)
Comment (mafredri) [reply coderd/database/queries/insights.sql:27]: I'll revert the 1m -> 5m change and I can work on query optimizations in another PR (optimization hasn't been a priority yet).

File: coderd/database/queries/insights.sql (hunk 3 above)
Comment (mafredri) [reply coderd/database/queries/insights.sql:34]: I've kept both at 5 minute roundups now for speed (trade-off for correctness).

File: coderd/insights.go (hunk 4 above)
Comment (mafredri) [reply coderd/insights.go:222]: I will create an improved test-suite as a follow-up to this PR which will perform tests for multiple templates as well.

**Comment by mafredri** [comment]
I've added basic tests and will improve upon them in a follow up PR. I've also manually confirmed that the query changes produces the same data on `dev.coder.com` when disregarding the new data.

A funny side-effect of using the 5 minute windows is that when testing, I:

1. Opened code-server
2. Refreshed insights, there was immediately 5 minutes code-server usage
3. Waited 1 minute
4. Refreshed insights, there was 10 minutes code-server usage because the clock went from :39 to :40

More users will skew nr. 4 even more, so I'm convinced we need to reduce the window whilst maintaining or improving query performance, and perhaps not use one at all for template apps since it's not strictly required.

**Review by johnstcn — APPROVED** [review]

**Review by mtojek — APPROVED** [review]
Is there anything left for @BrunoQuaresma to polish in the site area?
